### PR TITLE
Rerun suite, test, url

### PR DESCRIPTION
**Pull request by plutasnyy**
## Description
This PR provide you the functionality to rerun the whole suite - it runs in the same way like typical suite run, only test or url rerun - rerun part of suite work in a different way, after rerun, artifacts in your suite will be overwritten and aet doesn't create a new version of the suite.
The new way to run a test for page requires changes in the cleaner. The cleaner should remove overwritten artifacts- we don't want rubbish in our database. Functionality for the cleaner is in this PR

## Motivation and Context
It would be a nice improvement for users

## Screenshots (if appropriate):
44725131-e82f4200-aad4-11e8-9466-4d1994da8ca7
44725134-e9f90580-aad4-11e8-9b1a-59263d1b76ff
46950715-ec63ee80-d085-11e8-98db-9d7c609afb83
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](https://github.com/Cognifide/aet/blob/master/CONTRIBUTING.md#coding-conventions) of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

---
I hereby agree to the terms of the AET Contributor License Agreement.

### CODE REVIEW

File: api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/Comparator.java [hunk 1]
```diff
@@ -49,6 +50,9 @@ public void setStepResult(ComparatorStepResult stepResult) {
   }
 
   public List<Operation> getFilters() {
+    if(filters == null){
```
Comment (dlewandowski-cognifide) [inline api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/Comparator.java]: is it possible to pass the condition?

File: api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/Suite.java [hunk 2]
```diff
@@ -52,8 +54,12 @@
   private static final Type SUITE_TYPE = new TypeToken<Suite>() {
   }.getType();
 
+  public void setCorrelationId(String correlationId) {
```
Comment (dlewandowski-cognifide) [inline api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/Suite.java]: why do we have the setter so high in the class?

File: api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/Suite.java [hunk 3]
```diff
@@ -191,7 +208,7 @@ public String toString() {
         .toString();
   }
 
-  public void setVersion(long version) {
+  public void setVersion(Long version) {
```
Comment (dlewandowski-cognifide) [inline api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/Suite.java]: is it expected that `version` could be null?

Comment (plutasnyy) [reply api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/Suite.java]: Yes,
I remember that it is strongly connected with updating version of a suite. Suite version is null during the first run of and typical run but during rerun, the suite has a version, so I needed to allow on setting empty version and set the same state like in a typical run. I didn't see a better way to set the correct suite's version during the rerun.

File: api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/Suite.java (hunk 2 above)
Comment (plutasnyy) [reply api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/Suite.java]: Fixed 😄

File: api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/Test.java [hunk 4]
```diff
@@ -115,4 +116,25 @@ public String toString() {
...
+
+  public Optional<Url> getUrl(String urlName) {
+    Url urlToReturn = null;
+    for (Url url: this.urls) {
+      if(url.getName().equals(urlName)){
+        urlToReturn = url;
+        break;
+      }
+    }
+    return Optional.ofNullable(urlToReturn);
```
Comment (Skejven) [inline api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/Test.java]: This is just cosmetics and your implementation is perfectly fine. But you could also do this method body with streams:
```suggestion
    return urls.stream()
        .filter(url -> url.getName().equals(urlName))
        .findFirst();
```

File: core/runner/src/main/java/com/cognifide/aet/runner/processing/processors/SuiteExecutionProcessorStrategy.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+import com.cognifide.aet.communication.api.metadata.ValidatorException;
+import com.cognifide.aet.communication.api.wrappers.SuiteRunWrapper;
+import com.cognifide.aet.runner.processing.data.wrappers.RunIndexWrapperFactory;
+import com.cognifide.aet.vs.StorageException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SuiteExecutionProcessorStrategy extends ProcessorStrategy {
+
+  protected static final Logger logger = LoggerFactory
```
Comment (Skejven) [inline core/runner/src/main/java/com/cognifide/aet/runner/processing/processors/SuiteExecutionProcessorStrategy.java]: Name should be all upercase `LOGGER` a this is constant.

File: core/runner/src/main/java/com/cognifide/aet/runner/processing/processors/TestExecutionProcessorStrategy.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import com.cognifide.aet.communication.api.metadata.ValidatorException;
+import com.cognifide.aet.runner.processing.data.wrappers.RunIndexWrapperFactory;
+import com.cognifide.aet.vs.SimpleDBKey;
+import com.cognifide.aet.vs.StorageException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TestExecutionProcessorStrategy extends ProcessorStrategy {
+
+  protected static final Logger logger = LoggerFactory
```
Comment (Skejven) [inline core/runner/src/main/java/com/cognifide/aet/runner/processing/processors/TestExecutionProcessorStrategy.java]: Name should be all upercase `LOGGER` a this is constant.

File: core/runner/src/main/java/com/cognifide/aet/runner/processing/processors/UrlExecutionProcessorStrategy.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+import com.cognifide.aet.communication.api.metadata.ValidatorException;
+import com.cognifide.aet.runner.processing.data.wrappers.RunIndexWrapperFactory;
+import com.cognifide.aet.vs.SimpleDBKey;
+import com.cognifide.aet.vs.StorageException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UrlExecutionProcessorStrategy extends ProcessorStrategy {
+
+  protected static final Logger logger = LoggerFactory
```
Comment (Skejven) [inline core/runner/src/main/java/com/cognifide/aet/runner/processing/processors/UrlExecutionProcessorStrategy.java]: Name should be all upercase `LOGGER` a this is constant.

**Comment by plutasnyy** [comment]
Thank you for help with documentation and review

File: CHANGELOG.md [hunk 8]
```diff
@@ -18,6 +18,7 @@ All notable changes to AET will be documented in this file.
 - [PR-396](https://github.com/Cognifide/aet/pull/396) Added horizontal scrollbar for wide pages ([#393](https://github.com/Cognifide/aet/issues/393)) 
 - [PR-403](https://github.com/Cognifide/aet/pull/403) Conditionally passed tests can be accepted ([#400](https://github.com/Cognifide/aet/issues/400))
 - [PR-387](https://github.com/Cognifide/aet/pull/387) Set max allowed page screenshot height to 35k pixels.
+- [PR-412](https://github.com/Cognifide/aet/pull/412) ([PR-336](https://github.com/Cognifide/aet/pull/336), [PR-337](https://github.com/Cognifide/aet/pull/337), [PR-395](https://github.com/Cognifide/aet/pull/395)) - Added rerun functionality for suite, test and url
```
Comment (tkaik) [inline CHANGELOG.md]: @plutasnyy Please move this entry to "Unreleased" section

Comment (plutasnyy) [reply CHANGELOG.md]: Done 😃

**Comment by Slasheruus** [comment]
Tested (/)

**Pull request merged**